### PR TITLE
Only point out a single function parameter if we have a single arg incompatibility

### LINKED ISSUE #99635: Function signature probably shouldn't be reported if an argument type mismatch is 'trivial'
I've been recently doing a large amount of refactoring on my compiler. Although this has broadly been a pleasant experience, recent diagnostic changes have put a little pressure on my scroll wheel. Diagnostics that I (and I think many others, even newcomers) would consider trivial can sometimes take up many dozens of lines of compiler output. I definitely appreciate the verbosity, but I don't think it's always called for (at least, by default). As an example, I recently encountered the following error message:

```
error[E0308]: mismatched types
    --> analysis/src/infer.rs:1959:66
     |
1959 | ...                   .for_each(|(x, y)| self.derive_links(x, y, self_ty, ty_links, eff_links)),
     |                                               ------------ ^ expected struct `tao_util::index::Id`, found reference
     |                                               |
     |                                               arguments to this function are incorrect
     |
     = note: expected struct `tao_util::index::Id<(tao_syntax::Span, ty::Ty)>`
             found reference `&tao_util::index::Id<(tao_syntax::Span, ty::Ty)>`
note: associated function defined here
    --> analysis/src/infer.rs:1866:8
     |
1866 |     fn derive_links(
     |        ^^^^^^^^^^^^
1867 |         &mut self,
     |         ---------
1868 |         member: TyId,
     |         ------------
1869 |         ty: TyVar,
     |         ---------
1870 |         self_ty: Option<TyVar>,
     |         ----------------------
1871 |         ty_links: &mut HashMap<usize, TyVar>,
     |         ------------------------------------
1872 |         eff_links: &mut HashMap<usize, EffectVar>,
     |         -----------------------------------------
help: consider dereferencing the borrow
     |
1959 |                             .for_each(|(x, y)| self.derive_links(*x, y, self_ty, ty_links, eff_links)),
     |                                                                  +
```

That's 31 lines of terminal output to inform me that I forgot a `*`. As I say, I do appreciate the verbosity! But in cases like this, where the compiler can determine that the mismatch is trivial and the intended result is obvious (i.e: dereference the reference), I feel that the compiler should omit some of the more verbose parts of the message, such as the signature of the function.

I realise that this is, to some extent, bikeshedding: but there is a very real trade-off between the assistance provided by a compiler error and redundant information in the error message that makes the cycle of fixing error messages slower and more inconvenient.

In my view, an output like the following where the original signature has been omitted has a much better tradeoff between verbosity and the time required to read the message and get to the heart of what the compiler is suggesting:

```
error[E0308]: mismatched types
    --> analysis/src/infer.rs:1959:66
     |
1959 | ...                   .for_each(|(x, y)| self.derive_links(x, y, self_ty, ty_links, eff_links)),
     |                                               ------------ ^ expected struct `tao_util::index::Id`, found reference
     |                                               |
     |                                               arguments to this function are incorrect
     |
     = note: expected struct `tao_util::index::Id<(tao_syntax::Span, ty::Ty)>`
             found reference `&tao_util::index::Id<(tao_syntax::Span, ty::Ty)>`
help: consider dereferencing the borrow
     |
1959 |                             .for_each(|(x, y)| self.derive_links(*x, y, self_ty, ty_links, eff_links)),
     |                                                                  +
```

I am happy to make a PR for this myself, but I've not worked on rustc diagnostics before and may need pointing to the relevant code.

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes #99635

### CODE REVIEW

File: src/test/ui/argument-suggestions/too-long.stderr [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+   |         |
+   |         arguments to this function are incorrect
+   |
+note: associated function defined here
+  --> $DIR/too-long.rs:4:8
+   |
+LL |     fn foo(
+   |        ^^^
+...
+LL |         f: i32,
```
Comment (compiler-errors) [inline src/test/ui/argument-suggestions/too-long.stderr:15]: We could say "parameter defined here" if we have just one (or maybe just if it's multiline)

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs [hunk 2]
```diff
@@ -1786,6 +1789,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         err: &mut rustc_errors::DiagnosticBuilder<'tcx, rustc_errors::ErrorGuaranteed>,
         callable_def_id: Option<DefId>,
         callee_ty: Option<Ty<'tcx>>,
+        // A specific argument should be labeled, instead of all of them
+        expected_idx: Option<usize>,
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1793]: this only works if there's a single error in the function arguments, right? would it be possible to extend it to multiple arguments? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7a212f35d6f07a993c223f3e4662db39

Comment (estebank) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1793]: The further we move from certainty, the more we might mislead if we point at things too confidently. I feel like doing this only for single args for now is a good compromise while we come up with better heuristics. The suggestion being wrong is counteracted by pointing at the whole fn signature, but if we no longer do that then we're implicitly claiming infallibility 😅

File: compiler/rustc_parse/src/parser/expr.rs [hunk 3]
```diff
@@ -2225,7 +2225,7 @@ impl<'a> Parser<'a> {
                     attrs: attrs.into(),
                     ty,
                     pat,
-                    span: lo.to(this.token.span),
+                    span: lo.to(this.prev_token.span),
```
Comment (estebank) [inline compiler/rustc_parse/src/parser/expr.rs:2228]: This change means that we no longer include the comma in the param span, right? I don't know if that is a good call 🤔 

Maybe we can address the span issue on the other end, by creating a new span that is `param.pat.span.to(param.ty.map_or(param.pat.span, |ty| ty.span)`. I guess it depends on what we've used the param's span for elsewhere (even though there haven't been any visible changes in our test suite from this... I guess I'm fine either way :-/

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs [hunk 4]
```diff
@@ -1886,14 +1892,32 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-            for param in params {
-                spans.push_span_label(param.span, "");
+            for (idx, param) in params.into_iter().enumerate() {
+                if let Some(expected_idx) = expected_idx {
+                    if idx == expected_idx {
+                        spans.push_span_label(param.span, "");
+                    }
+                } else {
+                    spans.push_span_label(param.span, "");
+                }
```
Comment (estebank) [inline compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1905]:
```suggestion
            for (idx, param) in params.into_iter().enumerate()
                .filter(|(idx, _)| expected_idx.map_or(true, |expected| idx == expected))
            {
                spans.push_span_label(param.span, "");
```

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs (hunk 2 above)
Comment (jyn514) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:1793]: For context, I was hoping this change would make the awful error in https://mobile.twitter.com/joshuayn514/status/1557913836337786880 better.

Maybe we can somehow limit this to cases where the arguments are "close", for some definition of close? like, say, a function pointers where only the arguments differ? :P

(I don't actually have great general suggestions here but I would love that error to be smarter than it is now ...)

File: compiler/rustc_parse/src/parser/expr.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_parse/src/parser/expr.rs:2228]: I think nobody used it, evidenced by the lack of UI test changes. 

This makes it more consistent with  regular fn signatures' spans.

Comment (estebank) [reply compiler/rustc_parse/src/parser/expr.rs:2228]: That's fine, let's go ahead with that change then. We can revisit this if we ever need the param + comma (I'm sure, for removal suggestions) :)

**Comment by compiler-errors** [comment]
alleviating pressure on bors queue, since this is just diagnostics changes 

@bors rollup

**Pull request merged**